Schedule/api

cryogenian opened the pull request:
Whoa, that's big 

+ Squashed `impl` and `run`
+ Added scheduler stuff to `api` project, even `Scheduler` and `SchedulerModule`
+ Added scheduler stuff to `impl`, mostly mimiced destination stuff
+ Added local schedulers 
+ @djspiewak's module loading stuff is here too.

djspiewak commented on impl/src/main/scala/quasar/impl/Quasar.scala:
```diff
@@ -149,7 +160,7 @@ object Quasar extends Logging {
         resultRender))
 
-    } yield new Quasar(datasources, destinations, tables, sqlEvaluator, discovery, push)
+    } yield new Quasar(datasources, destinations, tables, sqlEvaluator, discovery, push, schedulers)
```
Super-minor nit: it would be nice if all the plugins were together in the parameter block. So `schedulers` would come after `destinations`.

djspiewak commented on impl/src/test/scala/quasar/impl/local/LocalSchedulerSpec.scala:
```diff
@@ -0,0 +1,332 @@
+import LocalSchedulerSpec._
+
+// This is hacky because I'm not sure how to define Timer[IO] that works x times faster than default one
+// So, there is a timer that works x times fiaster and have a clock that returns predefined time.
```
https://github.com/typelevel/cats-effect/blob/master/laws/shared/src/main/scala/cats/effect/laws/util/TestContext.scala

cryogenian replied:
Oh! Thanks

cryogenian commented on build.sbt:
```diff
@@ -273,24 +275,14 @@ lazy val impl = project
       "com.fasterxml.woodstox" % "woodstox-core" % "6.0.2"))
   .evictToLocal("FS2_JOB_PATH", "core")
-
-lazy val runp = (project in file("run"))
```
I squashed these two projects mostly because of ability to hide things like `DefaultDestinations` and other `DefaultStuff` behind `private[impl]`, didn't make it yet. But that's was the idea. This way changes in `impl` wouldn't be breaking and so on. 

`private[impl]` seems better then separate project for entry point (and we still have it in sdbe)

cryogenian commented on api/src/main/scala/quasar/api/scheduler/SchedulerModule.scala:
```diff
@@ -14,11 +14,20 @@
+
+import SchedulerError._
+
+trait SchedulerModule[F[_], I] {
```
Done

cryogenian commented on api/src/main/scala/quasar/api/scheduler/Scheduler.scala:
```diff
@@ -0,0 +1,32 @@
+
+import fs2.Stream
+
+trait Scheduler[F[_], I, C] {
```
:+1:

cryogenian commented on api/src/main/scala/quasar/api/scheduler/Schedulers.scala:
```diff
@@ -0,0 +1,35 @@
+  def replaceScheduler(i: I, ref: SchedulerRef[C]): F[Condition[SchedulerError[I, C]]]
+  def supportedTypes: F[Set[SchedulerType]]
+  def intentions: Stream[F, (I, II, CC)]
+  def schedulerOf(i: I): F[Either[SchedulerError[I, C], Scheduler[F, II, CC]]]
```
:+1:

cryogenian commented on impl/src/main/scala/quasar/impl/external/ExternalModule.scala:
```diff
@@ -0,0 +1,64 @@
+object ExternalModule {
+  final case class Datasource[F[_]](mod: DatasourceModule) extends ExternalModule[F]
+  final case class Destination[F[_]](mod: DestinationModule) extends ExternalModule[F]
+  final case class Scheduler[F[_]](mod: SchedulerModule[F, UUID]) extends ExternalModule[F]
```
Done

cryogenian commented on impl/src/main/scala/quasar/impl/local/LocalScheduler.scala:
```diff
@@ -0,0 +1,174 @@
+
+  import Period._
+
+  def apply[F[_]: ContextShift: ConcurrentEffect: Timer, I, C: EncodeJson: DecodeJson: Submit[F, ?]](
```
Done. Well not really done, I've removed local scheduler completely, as we don't need _this_ local scheduler in production and `ephemeral` is Ding an sich, isn't used anywhere except tests that check it works

cryogenian commented on api/src/main/scala/quasar/api/scheduler/SchedulerModule.scala:
```diff
@@ -14,11 +14,20 @@
+
+import SchedulerError._
+
+trait SchedulerModule[F[_], I] {
```
Wouldn't making `I` and `C` type members with contracts of decoding/encoding to jsons are just additional constraint to internal scheduler module w/o any gain? Concrete implementations could decide to use anything with any codecs and handle `UUID/Json` pair as they wish with or without `SchedulerModule` having type members.

wemrysi replied:
Making them type members here along with codecs allows `Scheduler` implementations to work with concrete types instead of serialization formats and allows (de)serialization to be defined/handled centrally in quasar, relieving the schedulers of that burden.

In any case, I think we'll want to use something more generic than `UUID` as the concrete type for `I`, like `ByteVector`. This allows scheduler implementations that already have an identifier format to use it directly without having to worry about embedding or translating UUIDs into native identifiers.

cryogenian commented:
@wemrysi This is ready for another look

+ Added `SchedulerModule` instead of `SchedulerModule[F[_], I]`
+ Moved `Scheduler` and `SchedulerModule` to `connector`
+ Added `Intentions` to `api`, `DefaultIntentions` to `impl`
+ Tests for intentions
+ Removed redundant type argument from `DefaultDestinations`
+ Made `SchedulerModule.schedule` returning `Scheduler[F, Array[Byte], Json]`

I tried  to make `I` and `C` type members, but it's a bit problematic because 
+ `I` and `C` are internal to `Scheduler[F]` because `Schedulers` should work with any `Scheduler[F]`
+ OTOH `Intensions` needs unified indices for `II`. But this is possible only when `DefaultIntentions` actually decodes/encodes `Array[Byte]` (`ByteVector`, whatever else) into internal `Scheduler.I` based on `lookup` function. 

This looks a bit ugly to me because it's not immediately clear why this conversions should hapen in `DefaultIntentions`. I'll add something like `xmapIndex` and `xmapConfig` in next pr if you don't mind. The error types are tricky there.

cryogenian commented:
Added `xmapIndex` and `xmapConfig` to `Schedule` trait.

cryogenian commented:
+ Moved packages `impl.store` and `impl.storage` are squished
+ `Schedulers` has ability of enabling and disabling `SchedulerModule` 

Put 🛑 because I want to add `pushDestination` and `cancelDestination` methods to `ResultPush` before merging

wemrysi commented:
@cryogenian would you mind moving the push changes to a separate PR? This one is already huge and it'd be nice to keep it as focused as possible on scheduling.

cryogenian commented:
@wemrysi Yes, no problem

cryogenian commented:
Removed push wip from this pr

wemrysi commented on impl/src/test/scala/quasar/impl/scheduler/DefaultSchedulersSpec.scala:
```diff
@@ -0,0 +1,245 @@
+        }
+      }
+    }
+    "replace destination" >> {
```
s/destination/scheduler
